libcudacxx/include/cuda/std/__cmath/min_max.h

```diff
@@ -21,224 +21,131 @@
-#  include <math.h>
-#endif // _CCCL_COMPILER(MSVC) || _CCCL_CUDA_COMPILER(CLANG)
-
 #include <cuda/std/__cccl/prologue.h>
 
 _LIBCUDACXX_BEGIN_NAMESPACE_STD
 
 // fmax
-
-#if _CCCL_CHECK_BUILTIN(builtin_fmax) || _CCCL_COMPILER(GCC)
-#  define _CCCL_BUILTIN_FMAXF(...) __builtin_fmaxf(__VA_ARGS__)
-#  define _CCCL_BUILTIN_FMAX(...)  __builtin_fmax(__VA_ARGS__)
-#  define _CCCL_BUILTIN_FMAXL(...) __builtin_fmaxl(__VA_ARGS__)
-#endif // _CCCL_CHECK_BUILTIN(builtin_fmax)
-
-[[nodiscard]] _CCCL_API inline float fmax(float __x, float __y) noexcept
-{
-#if defined(_CCCL_BUILTIN_FMAX)
-  return _CCCL_BUILTIN_FMAXF(__x, __y);
-#else // ^^^ _CCCL_BUILTIN_FMAX ^^^ / vvv !_CCCL_BUILTIN_FMAX vvv
-  return ::fmaxf(__x, __y);
-#endif // !_CCCL_BUILTIN_FMAX
-}
-
-[[nodiscard]] _CCCL_API inline float fmaxf(float __x, float __y) noexcept
-{
-#if defined(_CCCL_BUILTIN_FMAX)
-  return _CCCL_BUILTIN_FMAXF(__x, __y);
-#else // ^^^ _CCCL_BUILTIN_FMAX ^^^ / vvv !_CCCL_BUILTIN_FMAX vvv
-  return ::fmaxf(__x, __y);
-#endif // !_CCCL_BUILTIN_FMAX
-}
-
-[[nodiscard]] _CCCL_API inline double fmax(double __x, double __y) noexcept
-{
-#if defined(_CCCL_BUILTIN_FMAX)
-  return _CCCL_BUILTIN_FMAX(__x, __y);
-#else // ^^^ _CCCL_BUILTIN_FMAX ^^^ / vvv !_CCCL_BUILTIN_FMAX vvv
-  return ::fmax(__x, __y);
-#endif // !_CCCL_BUILTIN_FMAX
-}
-
-#if _CCCL_HAS_LONG_DOUBLE()
-[[nodiscard]] _CCCL_API inline long double fmax(long double __x, long double __y) noexcept
-{
-#  if defined(_CCCL_BUILTIN_FMAX)
-  return _CCCL_BUILTIN_FMAXL(__x, __y);
-#  else // ^^^ _CCCL_BUILTIN_FMAX ^^^ / vvv !_CCCL_BUILTIN_FMAX vvv
-  return ::fmaxl(__x, __y);
-#  endif // !_CCCL_BUILTIN_FMAX
-}
-[[nodiscard]] _CCCL_API inline long double fmaxl(long double __x, long double __y) noexcept
+_CCCL_TEMPLATE(class _Tp)
+_CCCL_REQUIRES(__is_extended_arithmetic_v<_Tp>)
+[[nodiscard]] _CCCL_API constexpr auto fmax(_Tp __x, _Tp __y) noexcept
 {
-#  if defined(_CCCL_BUILTIN_FMAX)
-  return _CCCL_BUILTIN_FMAXL(__x, __y);
-#  else // ^^^ _CCCL_BUILTIN_FMAX ^^^ / vvv !_CCCL_BUILTIN_FMAX vvv
-  return ::fmaxl(__x, __y);
-#  endif // !_CCCL_BUILTIN_FMAX
-}
-#endif // _CCCL_HAS_LONG_DOUBLE()
-
 #if _LIBCUDACXX_HAS_NVFP16()
-[[nodiscard]] _CCCL_API inline __half fmax(__half __x, __half __y) noexcept
-{
-  NV_IF_ELSE_TARGET(NV_IS_DEVICE,
-                    (return ::__hmax(__x, __y);),
-                    (return __float2half(_CUDA_VSTD::fmaxf(__half2float(__x), __half2float(__y)));))
-}
-template <class _A1, enable_if_t<is_arithmetic_v<_A1>, int> = 0>
-[[nodiscard]] _CCCL_API inline __promote_t<float, _A1> fmax(__half __x, _A1 __y) noexcept
-{
-  return _CUDA_VSTD::fmaxf(__half2float(__x), __y);
-}
-
-template <class _A1, enable_if_t<is_arithmetic_v<_A1>, int> = 0>
-[[nodiscard]] _CCCL_API inline __promote_t<_A1, float> fmax(_A1 __x, __half __y) noexcept
-{
-  return _CUDA_VSTD::fmaxf(__x, __half2float(__y));
-}
+  if constexpr (is_same_v<_Tp, __half>)
+  {
+    NV_IF_ELSE_TARGET(NV_IS_DEVICE,
```

Comment:
We historically did this dance with host device, but I am not sure why. We only allow it for 12.2 plus, so `__hmax` should be host device.

@fbusato thoughts?